### PR TITLE
Katrina - Edges - Hotel 

### PR DESCRIPTION
**Pull request by kaganon**
# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | In the beginning of my design I had a room class, which contained the room numbers and cost. However, I ended up removing the room class because I realized I that I could just keep track of all the rooms in my BookingSystem class, and that the cost of the room would be stored as a constant variable in my Reservation class. 
Describe a concept that you gained more clarity on as you worked on this assignment. | I think after drawing out what I needed from each class and seeing how the Booking System is the main driver code that connects all the other classes, it was a little bit clearer for me to figure out the direction of my program.
Describe a _nominal test_ that you wrote for this assignment. | A nominal test in my program include correctly calculating total cost of a reservation. It tests to make sure that it calculates the total cost of the reservation, even when adding multiple dates to the same reservation. 
Describe an _edge case test_ that you wrote for this assignment. | An edge case I wrote included raising an error for invalid dates. For example if the user entered nil for either or both dates, it would raise an error.
How do you feel you did in writing pseudocode first, then writing the tests and then the code? | I think it’s still challenging for me to think of pseudo code and tests first before writing code, especially if I’m not sure in my design of the code. But I’ve found that it’s been getting better with more practice.

### CODE REVIEW

File: lib/booking_system.rb [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+    def initialize
+      @rooms = (1..20).to_a
+      @reservations = []
+      @block_rooms = []
+    end
+
+    def list_rooms
+      return @rooms
+    end
```
Comment (droberts-sea) [inline lib/booking_system.rb]: This is doing the same thing as `attr_reader :rooms`, just with a different name. I would say that the `attr_reader` hits the project requirement, and you could omit this method.

File: lib/booking_system.rb [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+    def list_reservations(date_range)
+      return @reservations.select { |res| res.date_range == date_range }
+    end
```
Comment (droberts-sea) [inline lib/booking_system.rb:20]: Good use of `select` here.

File: lib/booking_system.rb [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    def list_available_rooms(date_range)
+      return @rooms if @reservations.empty?
+      return find_available_rooms(date_range)
+    end
+
+    def find_available_rooms(date_range)
+      return @rooms if @reservations.empty?
+      unavailable_rooms = @reservations.map { |res| res.room_number if date_range.dates_overlap?(res.date_range) }
+
+      unavailable_rooms_block = @block_rooms.map { |res| res.block_rooms if date_range.dates_overlap?(res.date_range)}.flatten
```
Comment (droberts-sea) [inline lib/booking_system.rb]: This code has some problems. On line 31, you'll end up with a bunch of `nil`s in your array, since `map` must always produce a value even if none of the code in the block executes.

On line 33, not only will you have `nil`, but the elements in your array are more arrays, since that's what a `BlockRoom` stores. That means that after line 35 has run, you'll have an array like:

```ruby
[nil, 3, nil, nil, 5, nil, [8, 9, 10], nil]
```

A side effect of this is that your code does not respect blocks when making reservations!

You could use a combination of `Array#flatten` and `Array#compact` to fix it, or it could be that `map` isn't a great choice for this problem.

File: lib/booking_system.rb [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+    def find_block(date_range)
+      raise StandardError, "No block rooms found for those dates" if @block_rooms.empty?
+      return @block_rooms.select { |block| block.date_range == date_range }
+    end
+
+    def find_room_in_block(date_range)
+      found_block = find_block(date_range)
+      room_block_rooms = found_block.first.block_rooms
+      available_rooms = find_available_rooms(date_range)
```
Comment (droberts-sea) [inline lib/booking_system.rb:72]: What if there's more than one block for a date range?

File: lib/booking_system.rb (hunk 3 above)
Comment (droberts-sea) [reply lib/booking_system.rb]: I do love the way this code flows though - logic problems aside, this is an excellent example of loose coupling. Rather than pulling the dates out of the reservation and doing a bunch of math here, you call a method and let the date range take care of it. That makes this code easier to read, and it means if something about how overlaps are calculated changes this code won't have to. Good work!

File: lib/date_range.rb [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    def initialize(start_date, end_date)
+      @start_date = Date.parse(start_date)
+      @end_date = Date.parse(end_date)
+    end
+
+    def check_valid_dates
+
+      raise StandardError, "Invalid dates entered" if @start_date == nil || @end_date == nil
+
+      raise StandardError, "End date cannot be before start date" if (@end_date - @start_date) <= 0
```
Comment (droberts-sea) [inline lib/date_range.rb]: Instead of raising a `StandardError`, best practice is to define your own appropriately named exception type that inherits from `StandardError` and raise that. The reason is that `StandardError` is very broad, and if the code that calls this method has to `rescue StandardError`, that will catch things like `ArgumentError`s and `NameError`s that indicate true bugs.

File: lib/date_range.rb [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+      booked_dates = [*date_range.start_date..date_range.end_date]
+      new_dates = [*@start_date..@end_date]
+
+      #TODO: refactor 
+
+      # completely containing
+      if new_dates.first < booked_dates.first && new_dates.last > booked_dates.last
+        return true
+        # back end
+      elsif new_dates.first < booked_dates.last && new_dates.last > booked_dates.last
```
Comment (droberts-sea) [inline lib/date_range.rb]: I like the way you've placed this logic on the `DateRange` class - good division of responsibilities.

However, this check can be done in a single line! I challenge you to figure out how. Or you can go read the instructor implementation.

File: spec/booking_system_spec.rb [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+    end
+
+    it 'Correctly returns an array of available rooms for the date range' do
+      expect(@reservation.find_available_rooms(@date_range_1)).must_equal [*14..20]
+    end
+
+    it 'Raises an error if there are no available rooms for the date range' do
+        hotel_booked = Hotel::BookingSystem.new()
+
+        20.times do
```
Comment (droberts-sea) [inline spec/booking_system_spec.rb:152]: Good work catching this error case!

File: spec/date_range_spec.rb [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+      }.must_raise StandardError
+
+    end
+  end
+
+
+  describe 'dates_overlap? method' do
+    before do
+      @date_range = @reservation_1.reservations[0].date_range
+    end
```
Comment (droberts-sea) [inline spec/date_range_spec.rb:62]: Though the list of these checks may originally have come from the instructors, you did the work of interpreting them and implementing them as tests. I appreciate your diligence.

**Comment by droberts-sea** [comment]
# Hotel
## What We're Looking For

Feature | Feedback
---     | ---
**Baseline** |
Used git regularly	| yes
Answer comprehension questions	| yes
**Design** |
Each class is responsible for a single piece of the program | yes
Classes are loosely coupled | yes
**Wave 1** |
List rooms | 
Reserve a room for a given date range | yes
List reservations for a given date | yes
Calculate reservation price | yes
Invalid date range produces an error | yes
Test coverage | yes
**Wave 2** |
View available rooms for a given date range | yes
Reserving a room that is not available produces an error | yes
Test coverage | yes
**Wave 3** |
Create a block of rooms | yes
Check if a block has rooms | yes
Reserve a room from a block | yes
Test coverage | yes
**Fundamentals** | 
Names variables, classes and modules appropriately | yes
Understanding of variable scope - local vs instance | yes
Can create complex logical structures utilizing variables | yes
Appropriately uses methods to break down tasks into smaller simpler tasks | yes
Understands the differences between class and instance methods | yes
Appropriately uses iterators and enumerables | mostly
Appropriately writes and utilizes classes | yes
Appropriately utilizes modules as a mechanism to encapsulate functionality | yes
**Wrap Up** |
There is a refactors.txt file | yes
The file provides a roadmap to future changes | yes
**Additional Feedback** | Great job overall! Your design is clean, your code well-written and your test coverage solid. I spotted at least one logic bug, and as always there are a few places where things could be cleaned up, but in general I am quite happy with what I see here. Keep up the hard work!
